examples/C02_PM_SHT_OLED_WIFI_INFLUX/C02_PM_SHT_OLED_WIFI_INFLUX.ino

```diff
@@ -0,0 +1,192 @@
+/*
+This is the code for the AirGradient DIY Air Quality Sensor with an ESP8266 Microcontroller.
+
+It is a high quality sensor showing PM2.5, CO2, Temperature and Humidity on a small display and can send data over Wifi.
+
+For build instructions please visit https://www.airgradient.com/diy/
+
+Compatible with the following sensors:
+Plantower PMS5003 (Fine Particle Sensor)
+SenseAir S8 (CO2 Sensor)
+SHT30/31 (Temperature/Humidity Sensor)
+
+Please install ESP8266 board manager (tested with version 3.0.0)
+
+The codes needs the following libraries installed:
+"WifiManager by tzapu, tablatronix" tested with Version 2.0.3-alpha
+"ESP8266 and ESP32 OLED driver for SSD1306 displays by ThingPulse, Fabrice Weinberg" tested with Version 4.1.0
+
+Configuration:
+Please set in the code below which sensor you are using and if you want to connect it to WiFi.
+
+If you are a school or university contact us for a free trial on the AirGradient platform.
+https://www.airgradient.com/schools/
+
+MIT License
+*/
+
+#include <AirGradient.h>
+#include <WiFiManager.h>
+#include <ESP8266WiFi.h>
+#include <ESP8266HTTPClient.h>
+#include <InfluxDbClient.h>
+
+#include "SSD1306Wire.h"
+
+AirGradient ag = AirGradient();
+
+SSD1306Wire display(0x3c, SDA, SCL);
+
+#define DISPLAY_DELAY 3000
+#define INFLUX_SEND_PERIOD 5 * 60 * 1000
+static unsigned long lastSampleTime = 0 - INFLUX_SEND_PERIOD;  // initialize such that a reading is due the first time through loop()
+
+
+// set sensors that you do not use to false
+boolean hasPM=true;
+boolean hasCO2=true;
+boolean hasSHT=true;
+
+// set to true if you want to connect to wifi. The display will show values only when the sensor has wifi connection
+boolean connectWIFI=true;
+
+// InfluxDB server url, e.g. http://192.168.1.48:8086 (don't use localhost, always server name or ip address)
+#define INFLUXDB_URL "hostname"
+// InfluxDB database name
+#define INFLUXDB_DB_NAME "db_name"
+#define INFLUXDB_USER "db_user"
+#define INFLUXDB_PASSWORD "db_password"
+#define HOSTNAME "air_quality"
+
+// Single InfluxDB instance
+InfluxDBClient client(INFLUXDB_URL, INFLUXDB_DB_NAME);
+
+void setup(){
+  Serial.begin(9600);
+
+  display.init();
+  display.flipScreenVertically();
+  showTextRectangle("Init", String(ESP.getChipId(),HEX),true);
+
+  if (hasPM) ag.PMS_Init();
+  if (hasCO2) ag.CO2_Init();
+  if (hasSHT) ag.TMP_RH_Init(0x44);
+
+  if (connectWIFI) connectToWifi();
+  client.setConnectionParamsV1(INFLUXDB_URL, INFLUXDB_DB_NAME, INFLUXDB_USER, INFLUXDB_PASSWORD);
+  delay(2000);
+}
+
+void loop(){
+  unsigned long now = millis();
+  const boolean shouldPushToInflux = now - lastSampleTime >= INFLUX_SEND_PERIOD;
+  if (now - lastSampleTime >= INFLUX_SEND_PERIOD)
+    lastSampleTime += INFLUX_SEND_PERIOD;
```

Comment:
Hi - looking at pushing metrics to influxDB and came across this.

FYI millis() will overflow to 0 after around 50 days - once this happens it looks like this will fall apart one `now` resets as `now - lastSampleTime` would result in a negative value.

I may be wrong :)

FYI https://www.arduino.cc/reference/en/language/functions/time/millis/

Reply:
@McJoppy That is a good point, I hadn't considered that.  Open to suggestions here, might just make it a counter and change the setting to a multiple of the main loop sleep.

Reply:
I'm about to try this logic

**Setup**

- test sendPeriod not 0
- init 'lastDividedWhole' as unsigned long

**Loop**

- set now to millis()
- divide now by sendPeriod and assign to unsigned long sendTest
- if sendTest != lastDividedWhole then period must have passed

Reply:
My take on sending to InfluxDB can be found at https://gist.github.com/McJoppy/aab9e1f9e2e43fdea61eabe729122635

I tend to hack stuff together and over complicate stuff. I've tried to keep the original functionality there but disabled by default.

Also beware of line 50. I'm not a fan of the default font so have added my own Roboto which provides better info for me.

My remote server supports v1 so currently no support for v2 too.

Hope this is useful :)

Reply:
I pushed up a commit removing the use of `millis`. https://github.com/airgradienthq/arduino/pull/15/commits/001e15bcbf9e560256fbc27fdb99c44c3e6eaf00